DuckDuckGo/BrowserChromeManager.swift

```diff
@@ -33,68 +33,94 @@ protocol BrowserChromeDelegate: class {
 class BrowserChromeManager: NSObject, UIScrollViewDelegate {
 
     struct Constants {
-
-        static let threshold: CGFloat = 60
-
+        static let threshold: CGFloat = 30
     }
 
     weak var delegate: BrowserChromeDelegate?
 
-    var dragging = false
     var hidden = false
-    var lastOffset: CGPoint?
-    var cumulative: CGFloat = 0
-
+    var dragging = false
+    var draggingStartPosY: CGFloat = 0
+    
     func scrollViewDidScroll(_ scrollView: UIScrollView) {
         guard dragging else { return }
+        guard canHideBars(for: scrollView) else { return }
+        
+        let isInBottomBounceArea = scrollView.contentOffset.y > scrollView.contentOffsetYAtBottom
+        guard isInBottomBounceArea == false else { return }
 
-        if let lastOffset = lastOffset {
-
-            let ydiff = lastOffset.y - scrollView.contentOffset.y
-
-            if ydiff == 0 || (cumulative < 0 && ydiff > 0) || (cumulative > 0 && ydiff < 0) {
-                cumulative = 0
-            }
-
-            cumulative += ydiff
-
-            if abs(cumulative) > Constants.threshold {
-                updateBars(ydiff < 0)
-            }
-
+        if scrollView.contentOffset.y - draggingStartPosY > Constants.threshold {
+            updateBars(shouldHide: true)
         }
-
-        lastOffset = scrollView.contentOffset
     }
 
     func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
         dragging = true
+        draggingStartPosY = scrollView.contentOffset.y
+    }
+
+    func scrollViewWillEndDragging(_ scrollView: UIScrollView, withVelocity velocity: CGPoint, targetContentOffset: UnsafeMutablePointer<CGPoint>) {
+        guard canHideBars(for: scrollView) else { return }
+        
+        let isInBottomBounceArea = scrollView.contentOffset.y > scrollView.contentOffsetYAtBottom
+        let startedFromVeryBottom = abs(draggingStartPosY - scrollView.contentOffsetYAtBottom) < 1
+        
+        if isInBottomBounceArea && startedFromVeryBottom {
+            updateBars(shouldHide: false)
+            // Fix for iPhone X issue: reaching bottom of the web page and then revealing
+            // bars by executing "bottom bounce" gesture, caused web view to re layout and
+            // cover bottom of the web page.
+            DispatchQueue.main.async { [weak scrollView] in
+                guard let scrollView = scrollView else { return }
```

Comment:
just use `scrollView?.setContentOffset...` and you don't need this guard

Reply:
You could force unwrap the scrollview at that point as it must be safe to use because the optional wouldn't get that far, but you're right a guard is probably cleaner than that.

Reply:
Yup, I got the same impression! Thanks :)